### PR TITLE
Modify tests that write files to use a temp dir

### PR DESCRIPTION
**Pull request by steven-johnson**
Add a utility to create a temp directory just for testing purposes, and
modify the relevant tests to only write there (rather than
current-directory). This provides a useful bottleneck for making tests
work reliably in sandboxed environments (e.g. Bazel).

### CODE REVIEW

**Comment by steven-johnson** [comment]
Gentle tuesday morning PTAL ping

File: test/common/halide_test_dirs.h [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <stdio.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#endif
+
+namespace Halide {
+namespace Internal {
```
Comment (abadams) [inline test/common/halide_test_dirs.h]: I don't think it's necessary for this to be in the Internal namespace, given that it's not in Halide.h

Comment (abadams) [inline test/common/halide_test_dirs.h]: Actually, I guess the other file-handling methods are in Internal, so it makes sense.

**Comment by abadams** [comment]
LGTM. Only not merging it myself in case you come to a different conclusion to me about whether it should be in Internal.

File: test/common/halide_test_dirs.h (hunk 1 above)
Comment (steven-johnson) [inline test/common/halide_test_dirs.h]: Yeah, mimicking the other file-handlers. Don't feel strongly about it though. Will leave as-is for now.

**Comment by dsharletg** [comment]
If I had to guess, I think the failure in this log is probably due to this PR: https://buildbot.halide-lang.org/master/builders/linux-64-gcc49-trunk/builds/2217/steps/make%20test_correctness_1/logs/stdio

**Comment by abadams** [comment]
See also:

https://buildbot.halide-lang.org/master/builders/win-32-trunk/builds/1327/steps/correctness_1/logs/stdio

**Comment by steven-johnson** [comment]
Is this happening regularly or flakily? I haven't seen it on local builds.

**Comment by abadams** [comment]
The linux issue seems to be flaky, not consistent.

**Comment by abadams** [comment]
I think the windows problem was just a slash vs backslash thing. Pushed a
hopeful fix.

**Comment by steven-johnson** [comment]
Re: linux being flaky... if Travis is running several of various builds/tests in the filesystem at the same time, different tests could compete for the same "/tmp/halide_test" directory and cause this. We really need a unique temp directory for each test run. (Bazel makes this easy...)

**Comment by abadams** [comment]
This is not travis, it's the buildbots. They use separate directories per
builder for the builds, but they do run multiple different builds together.

**Comment by abadams** [comment]
(Can't resist a dig at bazel)

Before this change, made to appease bazel, we already had a separate temp
directory per build - bin/build/tmp within the working directory of the
builder.

**Comment by abadams** [comment]
Random thought: Should we jam the PID into the temporary dir name? That
should disambiguate multiple processes running concurrently.

**Comment by steven-johnson** [comment]
On Thu, Jan 19, 2017 at 2:49 PM, Andrew Adams <notifications@github.com>
wrote:

> (Can't resist a dig at bazel)
>
> Before this change, made to appease bazel, we already had a separate temp
> directory per build - bin/build/tmp within the working directory of the
> builder.
>

Yeahyeahyeah. I'm going to go back to using that, but in a way that will
appease Bazel.

ALL HAIL BAZEL

**Comment by steven-johnson** [comment]
See https://github.com/halide/Halide/pull/1763 for hopeful fix (pending testing)

**Pull request merged**